sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/UnusedFunctionParameterCheck.kt

```diff
@@ -19,54 +19,84 @@
 import org.sonarsource.kotlin.plugin.KotlinFileContext
 
 @Rule(key = "S1172")
 class UnusedFunctionParameterCheck : AbstractCheck() {
 
-    companion object {
-        private fun KtNamedFunction.getUnusedParameters(): List<KtParameter> =
-            valueParameters.asSequence()
-                .filter { ktParameter ->
-                    !anyDescendantOfType<KtNameReferenceExpression> {
-                        it.text == ktParameter.nameIdentifier?.text
-                    }
-                }.toList()
-    }
-
     override fun visitNamedFunction(function: KtNamedFunction, context: KotlinFileContext) {
         if (!shouldBeChecked(function)) return
         val unusedParameters = function.getUnusedParameters()
-        if (unusedParameters.isNotEmpty()) reportUnusedParameters(context, unusedParameters)
+        val singleMessage =
+            if (function.isAnonymous()) """Use "_" instead of this unused function parameter"""
+            else "Remove this unused function parameter"
+        val pluralMessage =
+            if (function.isAnonymous()) """Use "_" instead of these unused function parameters."""
+            else "Remove these unused function parameters."
+        if (unusedParameters.isNotEmpty()) reportUnusedParameters(context, unusedParameters, singleMessage, pluralMessage)
+    }
+
+    override fun visitLambdaExpression(expression: KtLambdaExpression, context: KotlinFileContext) {
+        val unusedParameters = expression.getUnusedParameters()
+        val singleMessage = """Use "_" instead of this unused lambda parameter"""
+        val pluralMessage = """Use "_" instead of these unused lambda parameters."""
+        if (unusedParameters.isNotEmpty()) reportUnusedParameters(context, unusedParameters, singleMessage, pluralMessage)
     }
 
-    private fun reportUnusedParameters(context: KotlinFileContext, unusedParameters: List<KtParameter>) {
+    private fun reportUnusedParameters(
+        context: KotlinFileContext,
+        unusedParameters: List<KtNamedDeclaration>,
+        singleMessage: String,
+        pluralMessage: String,
+    ) {
         val secondaryLocations = unusedParameters.asSequence()
-            .map { unusedParameter: KtParameter ->
+            .map { unusedParameter: KtNamedDeclaration ->
                 SecondaryLocation(context.textRange(unusedParameter.nameIdentifier!!),
-                    "Remove this unused method parameter ${unusedParameter.name}\".")
+                    """$singleMessage "${unusedParameter.name}".""")
             }
             .toList()
         val firstUnused = unusedParameters[0]
-        val msg = if (unusedParameters.size > 1) {
-            "Remove these unused function parameters."
-        } else {
-            "Remove this unused function parameter \"${firstUnused.name}\"."
-        }
+        val msg = if (unusedParameters.size > 1) pluralMessage
+        else """$singleMessage "${firstUnused.name}"."""
```

Comment:
Currently, we raise the same message for the same location twice when we have only 1 unused parameter. I don't think that's valuable and it may even look like a bug, as the UI shows the same thing twice. Can we get rid of the secondary location in case there is only 1 unused parameter?

Reply:
Yep, I was thinking about it. We decided to leave it previously, to be consistent with what we had in SLang, but currently, it makes sense to fix this.